Spigot-Server-Patches/0418-Ticket-level-per-player-view-distance-API.patch

```diff
@@ -0,0 +1,212 @@
+                                 ChunkMapDistance.this.l.add(i);
+                             } else {
+@@ -412,10 +418,14 @@ public abstract class ChunkMapDistance {
+         public void a() {
+             super.a();
+             if (!this.g.isEmpty()) {
+-                LongIterator longiterator = this.g.iterator();
++                //Paper start ticket level pp-view-dist
++                ObjectIterator<Long2ByteMap.Entry> longiterator = this.getUpdateIterator();
+ 
+                 while (longiterator.hasNext()) {
+-                    long i = longiterator.nextLong();
++                    Long2ByteMap.Entry e = longiterator.next();
++                    long i = e.getLongKey();
++                    int vd = e.getByteValue();
++                    //Paper end
+                     int j = this.f.get(i);
+                     int k = this.c(i);
+ 
+@@ -430,7 +440,7 @@ public abstract class ChunkMapDistance {
+                             }
+ 
+                         });
+-                        this.a(i, k, this.c(j), this.c(k));
++                        this.a(i, k, this.c(j), k <= vd - 2, vd); //Paper ticket level pp-view-dist
+                     }
+                 }
+ 
+@@ -448,6 +458,7 @@ public abstract class ChunkMapDistance {
+ 
+         protected final Long2ByteMap a = new Long2ByteOpenHashMap();
+         protected final int b;
++        protected int pvd; //Paper ticket level pp-view-dist
+ 
+         protected b(int i) {
+             super(i + 2, 16, 256);
+@@ -455,6 +466,8 @@ public abstract class ChunkMapDistance {
+             this.a.defaultReturnValue((byte) (i + 2));
+         }
+ 
++        protected int getChunksMapping(long i) {return this.c(i);} //Paper OBFHELPER
++
+         @Override
+         protected int c(long i) {
+             return this.a.get(i);
+@@ -464,16 +477,34 @@ public abstract class ChunkMapDistance {
+         protected void a(long i, int j) {
+             byte b0;
+ 
++            //Paper start ticket level pp-view-dist
++            if(j == 0) {
++                ObjectSet<EntityPlayer> c = ChunkMapDistance.this.getChunkPlayers(i);
++                if (c != null && !c.isEmpty()) {
++                    ObjectIterator<EntityPlayer> pl = c.iterator();
++                    int[] vd = new int[c.size()];
++                    int x = 0;
++                    while (pl.hasNext()) {
++                        EntityPlayer p = pl.next();
++                        vd[x] = p.getViewDistance();
++                        java.util.Arrays.sort(vd);
++                        this.pvd = vd[c.size() - 1];
++                        x++;
++                    }
++                }
++            }
++            //Paper end
++
+             if (j > this.b) {
+                 b0 = this.a.remove(i);
+             } else {
+                 b0 = this.a.put(i, (byte) j);
+             }
+ 
+-            this.a(i, b0, j);
++            this.a(i, b0, j, (byte) this.pvd); //Paper ticket level pp-view-dist
```

Comment:
there's no reason for this, this map is the map for mob spawning

Reply:
what? this is a crucial section of the code, this line allows the per view distance values to actually be passed on and attached to the queued update chunks.